service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/BlobCopier.java

```diff
@@ -0,0 +1,110 @@
+        storageAccessService.buildBlobContainerClient(sourceContainer, sourceStorageAccount);
+    var destinationBlobContainerClient =
+        storageAccessService.buildBlobContainerClient(
+            destinationContainer, destinationStorageAccount);
+
+    // directories are presented as zero-length blobs, filter these out as they are
+    // not copy-able
+    var blobItems =
+        sourceBlobContainerClient.listBlobs().stream()
+            .filter(blobItem -> blobItem.getProperties().getContentLength() > 0);
+
+    var blobPollers =
+        blobItems.map(
+            sourceBlobItem ->
+                beginCopyBlob(
+                    storageAccessService,
+                    sourceContainer,
+                    sourceStorageAccount,
+                    userRequest,
+                    sourceBlobItem,
+                    sourceBlobContainerClient,
+                    destinationBlobContainerClient));
+
+    logger.info(
+        "Copying blobs [source_container_id = {}, source_workspace_id = {}, destination_container_id = {}, destination_workspace_id={}]",
+        sourceContainer.getResourceId(),
+        sourceContainer.getWorkspaceId(),
+        destinationContainer.getResourceId(),
+        destinationContainer.getWorkspaceId());
+    var pollResults = blobPollers.map(blobPoller -> blobPoller.waitForCompletion(MAX_POLL_TIMEOUT));
+    logger.info(
+        "Finished copying blobs [source_container_id = {}, source_workspace_id = {}, destination_container_id = {}, destination_workspace_id={}]",
+        sourceContainer.getResourceId(),
+        sourceContainer.getWorkspaceId(),
+        destinationContainer.getResourceId(),
+        destinationContainer.getWorkspaceId());
+
+    var resultsByStatus = pollResults.collect(groupingBy(PollResponse::getStatus));
+    return new BlobCopierResult(resultsByStatus);
+  }
+
+  private SyncPoller<BlobCopyInfo, Void> beginCopyBlob(
+      AzureStorageAccessService storageAccessService,
+      ControlledAzureStorageContainerResource sourceContainer,
+      ControlledAzureStorageResource sourceStorageAccount,
+      AuthenticatedUserRequest userRequest,
+      BlobItem sourceBlobItem,
+      BlobContainerClient sourceBlobContainerClient,
+      BlobContainerClient destBlobContainerClient) {
+    AzureSasBundle sasBundle =
+        storageAccessService.createAzureStorageContainerSasToken(
+            sourceContainer.getWorkspaceId(),
+            sourceContainer,
+            sourceStorageAccount,
+            userRequest,
+            null,
+            sourceBlobItem.getName(),
+            "r");
+    var sourceBlobClient = sourceBlobContainerClient.getBlobClient(sourceBlobItem.getName());
+    var destinationBlobClient = destBlobContainerClient.getBlobClient(sourceBlobItem.getName());
+    var sourceBlobUrl = sourceBlobClient.getBlobUrl() + "?" + sasBundle.sasToken();
+
+    return destinationBlobClient.beginCopy(sourceBlobUrl, Duration.ofSeconds(1));
```

Comment:
Why the 1 second duration?

Reply:
No good reason; This controls the interval between poll. According to the docs the default is 1 second anyway, so I'll remove.